transports/quic/src/connection.rs

```diff
@@ -0,0 +1,836 @@
+                match connection.finish(*id) {
+                    Ok(()) => {}
+                    Err(error) => warn!("Finishing stream {:?} failed: {}", id, error),
+                }
+            }
+        }
+        Poll::Pending
+    }
+}
+
+#[derive(Debug)]
+pub struct Upgrade {
+    muxer: Option<QuicMuxer>,
+}
+
+#[cfg(test)]
+impl Drop for Upgrade {
+    fn drop(&mut self) {
+        debug!("dropping upgrade!");
+        assert!(
+            self.muxer.is_none(),
+            "dropped before being polled to completion"
+        );
+    }
+}
+
+impl Future for Upgrade {
+    type Output = Result<(libp2p_core::PeerId, QuicMuxer), Error>;
+    fn poll(self: Pin<&mut Self>, cx: &mut Context) -> Poll<Self::Output> {
+        let muxer = &mut self.get_mut().muxer;
+        trace!("outbound polling!");
+        let res = {
+            let mut inner = muxer.as_mut().expect("polled after yielding Ready").inner();
+            if inner.connection.is_drained() {
+                return Poll::Ready(Err(Error::ConnectionError(
+                    inner
+                        .close_reason
+                        .clone()
+                        .expect("closed connections always have a reason; qed"),
+                )));
+            } else if inner.connection.is_handshaking() {
+                assert!(
+                    !inner.connection.is_closed(),
+                    "a closed connection cannot be handshaking; qed"
+                );
+                inner.handshake_or_accept_waker = Some(cx.waker().clone());
+                return Poll::Pending;
+            } else if inner.connection.side().is_server() {
+                let endpoint = &mut inner.endpoint;
+                let token = ready!(endpoint.poll_ready(cx));
+                endpoint.send_message(EndpointMessage::ConnectionAccepted, token)
+            }
+
+            let peer_certificates: Vec<rustls::Certificate> = inner
+                .connection
+                .crypto_session()
+                .get_peer_certificates()
+                .expect("we have finished handshaking, so we have exactly one certificate; qed");
+            certificate::extract_libp2p_peerid(
+                peer_certificates
+                    .get(0)
+                    .expect(
+                        "our certificate verifiers require exactly one \
+                         certificate to be presented, so an empty certificate \
+                         chain would already have been rejected; qed",
+                    )
+                    .as_ref(),
+            )
+        };
+        let muxer = muxer.take().expect("polled after yielding Ready");
+        Poll::Ready(match res {
+            Ok(e) => Ok((e, muxer)),
+            Err(_) => Err(Error::BadCertificate(ring::error::Unspecified)),
+        })
+    }
+}
+
+type StreamSenderQueue = std::collections::VecDeque<oneshot::Sender<StreamId>>;
+
+#[derive(Debug)]
+pub(crate) struct Muxer {
```

Comment:
This entire module really needs some documentation so that programmers other than you (or you in a couple of weeks) can modify it later without having to spend half a day understanding the logic.

For example `pending_stream: Option<StreamId>` documented as "the pending stream, if any" isn't useful. I have that information from the field name and type already.
I don't understand what "pending" or "a stream" even mean. Are these incoming substreams?
This is the same for most of these fields. They all need to be documented in order to know in which context they are used.

Reply:
It absolutely does need documentation.  I wrote some, but it may need more.